tmt/utils.py

```diff
@@ -2360,19 +2360,22 @@ def web_git_url(url: str, ref: str, path: Optional[Path] = None) -> str:
     return url
 
 
+def run_cmd(command: str, cwd: Path) -> str:
```

Comment:
I'd recommend leaving this function inside `fmf_id` - moving it out suggests other parts of tmt are free to use it, which should not be encouraged as it lacks logging or exit code checking. We shouldn't tempt tmt codebase into using it instead of `self.run()` inherited from `tmt.utils.Common`.

Reply:
Hi @happz, in function `default_branch`, we also need to run command (the same as `run()` in function `fmf_id`). Any suggestion if we don't want to have duplicated code?

Reply:
I see. The best way, according to me? That would be liberating `Common._run()`, to become a standalone function we could then call from `Common._run()`, `fmf_id()` or `default_branch()`. But that would be more work, worth its own PR, TBH - I could help with that if you're not in a rush. On the other hand, we'd have just one way to "run commands", instead of two, and all commands would go through the same process, and have access to all goodies the `Common.run()` offers, like timeouts or environment setting.

Reply:
`Common.run()` could be usable now for this use case

Reply:
`Common` has at least some hidden dependencies, `Common.run()` and then `Common._run()` use `self.debug` and `self.workdir`. And mayeb something more, not 100% sure right now. `fmf_id()` and `default_branch()` would have to create `Common` instance just to call its `run()`, and then I find a standalone "run command" function with no implicit dependency on workdir or logger cleaner. `Common.run()` would become a thin wrapper to call the "run command" helper with `logger=self._logger, cwd=cwd or self.workdir` & get the original behavior.

Reply:
We do want to have a lightweight function to run shell command for this case, and the implement of `Common._run()` looks heavy, e.g.
* https://github.com/teemtee/tmt/blob/main/tmt/utils.py#L761
```python
   760	        # Run the command in interactive mode if requested
   761	        if interactive:
   762	            try:
   763	                subprocess.run(
   764	                    command.to_popen(),
   765	                    cwd=cwd, shell=shell, env=environment, check=True,
   766	                    executable=executable)
   767	            except subprocess.CalledProcessError:
   768	                # Interactive mode can return non-zero if the last command
   769	                # failed, ignore errors here
   770	                pass
   771	            finally:
   772	                return CommandOutput(None, None)
   773	
   774	        # Create the process
   775	        try:
   776	            process = subprocess.Popen(
   777	                command.to_popen(),
   778	                cwd=cwd, shell=shell, env=environment,
   779	                start_new_session=True,
   780	                stdin=subprocess.DEVNULL, stdout=subprocess.PIPE,
   781	                stderr=subprocess.STDOUT if join else subprocess.PIPE,
   782	                executable=executable)
   783	        except FileNotFoundError as error:
   784	            raise RunError(
   785	                f"File '{error.filename}' not found.",
   786	                command,
   787	                127,
   788	                caller=self)
```

> I find a standalone "run command" function with no implicit dependency on workdir or logger cleaner. 

Hi @happz , do you mean we need to add a standalone  function `run_command` to `Common` and call it in `fmf_id` and `default_branch`?

Reply:
> We do want to have a lightweight function to run shell command for this case, and the implement of `Common._run()` looks heavy

I agree, the implementation of `Common._run()` covers way more than is needed for a simple command. On the other hand, having 2 implementations introduce a question, when to use each of them? From the developer's point of view, having a single implementation is simply easier to navigate, as there's just one way to do "a thing". One does not have to worry, whether the command is simple enough, or whether its context is simple enough as well, and so on. Also, the implementation itself does not have to worry the caller of `_run()` - yeah, it can handle interactive commands, but that's a purely optional feature. On the other hand, one gets simplicity, consistency, and life-long maintenance for free :) With two implementations, it may happen we'd fix a bug in just one of them, for example :)

> > I find a standalone "run command" function with no implicit dependency on workdir or logger cleaner.
> 
> Hi @happz , do you mean we need to add a standalone function `run_command` to `Common` and call it in `fmf_id` and `default_branch`?

Yep, exactly that would be my approach. Here's the barely tested draft, https://github.com/teemtee/tmt/pull/1912

Reply:
@happz, thank you for providing a standalone `run_command()`, will have a try and get back to you.

Reply:
Hi @happz, I have updated the patch by using your `run_command`, and I manually tested it, the results looks good.  Please help to review, thanks!

```python
(tmt) huanli@kvm-01-guest02:TREE$ git remote -v
origin	https://github.com/teemtee/tmt (fetch)
origin	https://github.com/teemtee/tmt (push)
(tmt) huanli@kvm-01-guest02:TREE$ git branch 
+ main
* quay
(tmt) huanli@kvm-01-guest02:TREE$ python3 -m pdb ~/.virtualenvs/tmt/bin/tmt plans show -vvv
...<snip>..
(Pdb) b tmt/utils.py:2628
Breakpoint 1 at /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py:2628
(Pdb) c
/plans/features/advanced
...<snip>...
                 enabled true
                 sources /home/huanli/dev/1861/foo/REPRO/TREE/plans/main.fmf
                         /home/huanli/dev/1861/foo/REPRO/TREE/plans/features/advanced.fmf
> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py(2628)default_branch()
-> command = Command("git", "rev-parse", "--abbrev-ref", "HEAD")
(Pdb) args
repository = Path('/home/huanli/dev/1861/foo/REPRO/TREE')
logger = <Logger: name=tmt.logger0 verbosity=3 debug=0 quiet=False>
remote = 'origin'
(Pdb) n
> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py(2629)default_branch()
-> result = run_command(command=command, cwd=Path(git_parent_path), logger=logger)
(Pdb) p command
<tmt.utils.Command object at 0x7fd4d703d720>
(Pdb) n
> /home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/utils.py(2630)default_branch()
-> if result.stdout is None:
(Pdb) p result.stdout
'main\n'
```